### ISSUE TITLE
opus: seeking to wrong location and reading 0 frames?

### ISSUE THREAD

**Issue by krisfed** [libsndfile/libsndfile#794]
Hi all,

We are using libsndfile 1.0.30, and noticed an issue when trying to do partial reads of opus files. For **some specific indices** the sf_seek returns the number of frames that is different from requested, and the subsequent sf_readf_float returns 0 frames (even though it is not at the end of the file). For other indices within the file, the same workflow seems to work fine.
 
We observe that the "bad indices" are different for different files - there does not seem to be a clear pattern.

Here is some simple reproduction code:

```C++
#include <iostream>
#include "sndfile.h"

int main()
{
    // open file
    SF_INFO fileInfo;
    
    SNDFILE* file_sf = sf_open("myfile.opus", SFM_READ, &fileInfo);
    
    // seek
    sf_count_t startSample = 115502;
    sf_count_t framesSeeked = sf_seek(file_sf, startSample, SEEK_SET);
    
    std::cout << "startSample: " << startSample << std::endl;
    std::cout << "framesSeeked: " << framesSeeked << std::endl;
    
    // read
    sf_count_t framesRead = 0;
    
    size_t numSamples = 65536;
    size_t sizeInBytes = 4 * numSamples * 2;// type size * num samples * num channels
    uint8_t* buffer = new uint8_t[sizeInBytes];
    
    framesRead = sf_readf_float(file_sf, reinterpret_cast<float*>(buffer), numSamples);
    
    std::cout << "numSamples: " << numSamples << std::endl;
    std::cout << "sizeInBytes: " << sizeInBytes << std::endl;
    std::cout << "framesRead: " << framesRead << std::endl;
    
    sf_close(file_sf);
    
    return 0;
}
```

The output (note that "framesSeeked" is different from requested "startSample", and that 0 frames are read) :
```
startSample: 115502
framesSeeked: 94728
numSamples: 65536
sizeInBytes: 524288
framesRead: 0
```
The used opus file is in the attached [myfile.opus.zip](https://github.com/libsndfile/libsndfile/files/7590238/myfile.opus.zip). It has randomly generated data, and was written using libsndfile. 

The "bad indices" for this file seem to be in these (inclusive) ranges:

115502 - 146569
211704 - 242569
307906 - 338569

Other indices seem to work ok.

Here is the output of sndfile-info on this file:

```
$ sndfile-info myfile.opus 
========================================
File : myfile.opus
Length : 487995
Ogg stream data : Opus
Stream serialno : 2088749492
Opus library version: libopus unknown
Opus Header Metadata
  OggOpus version  : 1
  Channels		 : 2
  Preskip		  : 312 samples @48kHz
  Input Samplerate : 48000 Hz
  Gain			 : 0.0
  Channel Mapping  : 0 (mono or stereo)
VorbisComment Metadata
  Vendor: libopus unknown
  TITLE=Opus Quality default Fs 48000 Stereo
  ARTIST=Random generator

----------------------------------------
Sample Rate : 48000
Frames      : 480000
Channels    : 2
Format      : 0x00200064
Sections    : 1
Seekable    : TRUE
Duration    : 00:00:10.000
Signal Max  : 1.6254 (-86.09 dB)
```
But we saw the same issue on other files we created, including ones with real audio data.

I have also tried out the above repro code with libsndfile 1.0.31, and it behaves the same.

Does this look like a bug? Has anyone run into similar issues?

**Comment by evpobr** [comment]
Hello @krisfed . Thank you for the very useful information. I think @arthurt  will be interested to take a look at this.

**Comment by krisfed** [comment]
I wonder if this might be related to https://github.com/libsndfile/libsndfile/pull/556 ? Would appreciate any other info too!

**Comment by arthurt** [comment]
FYI: investigating.

**Comment by arthurt** [comment]
Dump off the sndfile log from opening the file
```
File : myfile.opus
Length : 487995
Ogg stream data : Opus
Stream serialno : 2088749492
Opus library version: libopus 1.3.1
Opus Header Metadata
  OggOpus version  : 1
  Channels         : 2
  Preskip          : 312 samples @48kHz
  Input Samplerate : 48000 Hz
  Gain             : 0.0
  Channel Mapping  : 0 (mono or stereo)
VorbisComment Metadata
  Vendor: libopus unknown
  TITLE=Opus Quality default Fs 48000 Stereo
  ARTIST=Random generator
  Granule pos offset  : 0
  Last Granule pos : 480312
```

Dump of the Ogg Stream structure:
```
Pg #	Offset  	Len	Flags	Pkts	GP        	GP Delta
=========================================================================
   0	00000000	    47	BOS    	  1	       0	       0
   1	00000047	   899	       	  1	       0	       0
   2	00000946	 47660	       	 49	   47040	   47040
   3	00048606	 48627	       	 50	   95040	   48000
   4	00097233	 48627	       	 50	  143040	   48000
   5	00145860	 48627	       	 50	  191040	   48000
   6	00194487	 48627	       	 50	  239040	   48000
   7	00243114	 48627	       	 50	  287040	   48000
   8	00291741	 48629	       	 50	  335040	   48000
   9	00340370	 48627	       	 50	  383040	   48000
  10	00388997	 48633	       	 50	  431040	   48000
  11	00437630	  4893	       	  5	  435840	    4800
  12	00442523	  4901	       	  5	  440640	    4800
  13	00447424	  4883	       	  5	  445440	    4800
  14	00452307	  4879	       	  5	  450240	    4800
  15	00457186	  4889	       	  5	  455040	    4800
  16	00462075	  4887	       	  5	  459840	    4800
  17	00466962	  4887	       	  5	  464640	    4800
  18	00471849	  4887	       	  5	  469440	    4800
  19	00476736	  4887	       	  5	  474240	    4800
  20	00481623	  4887	       	  5	  479040	    4800
  21	00486510	  1485	    EOS	  2	  480312	    1272
```

**Comment by arthurt** [comment]
Okay, made progress.

Ogg based formats have no seek-tables, and require file searching and page syncs to seek. The routine that does the seek position estimating and binary search of the Ogg pages is based on a routine from libopusfile.

Turns out the libopusfile routine has a bug. Need more parasitic data to test with, but I'll hopefully have something up soon. Then I guess I'll file a bug with libopusfile so they can fix the original routine as well.

This also impacts Vorbis seeking, although Vorbis is less likely to have an issue, having larger packets sizes of longer playback time.

On top of that, the "pre-roll" amount, that is the time it takes for decoder to converge exactly, that we estimated is wrong. It seems to depend on Opus packet sizes.

**Comment by arthurt** [comment]
Tested with parasitic data (Ogg pages that continue, packets that span multiple pages), and yes the fix is correct.

Fixes Vorbis seeking. Opus seeking still has an over count issue in near seeks. Just need to thing more about it.

**Comment by arthurt** [comment]
This issue, Opus seeking, has been fixed by PR #826.

Vorbis seeking, while better still experiences some issues on near-seeks.

**Comment by arthurt** [comment]
Closing as fixed. @krisfed please re-open if not :)
